shared_client.go

```diff
@@ -143,21 +144,44 @@ func handler(wg *sync.WaitGroup, client *Client, conn *Conn, requests chan reque
 
 	responses := make(chan sharedClientResponse)
 
+	receiverTrigger := make(chan struct{})
+	triggerReceiver := func() {
+		select {
+		case receiverTrigger <- struct{}{}:
+		default:
+		}
+	}
+
 	// Receive loop
 	wg.Add(1)
 	go func() {
 		defer wg.Done()
 		defer close(responses)
 		for {
+			// This will block but eventually return an i/o timeout, as we always set
+			// the timeouts before sending anything
 			r, err := conn.ReadMsg()
-			if err != nil {
-				// handler is not reading on the channel after closing
-				if errors.Is(err, net.ErrClosed) {
-					return
-				}
-				responses <- sharedClientResponse{nil, 0, err}
-			} else {
+			if err == nil {
 				responses <- sharedClientResponse{r, 0, nil}
+				continue // receive immediately again
+			}
+
+			// handler is not reading on the channel after closing.
+			// UDP connections return net.ErrClosed, while TCP/TLS connections are read
+			// via the io package, which return io.EOF.
+			if errors.Is(err, net.ErrClosed) || errors.Is(err, io.EOF) {
+				return
+			}
+
+			// send error response to cancel all current requests.
+			responses <- sharedClientResponse{nil, 0, err}
+
+			// wait for a trigger from the handler after any errors. Re-reading in
+			// this condition could busy loop, e.g., if a read timeout occurred.
+			_, ok := <-receiverTrigger
```

Comment:
Isn't this racy now?

If there is a trigger while we are failing in `ReadMsg`, the trigger can be missed even though subsequent read would succeed. 
Let's say there are two requests. We read on the conn, and the first read fails with `buffer too short`. Isn't it possible that the second trigger occurs during the (failing) read of the first response, and hence we wait even though there may be a valid response available?

Reply:
You may be right, but if the first error happens after the 2nd request has been sent, then also the 2nd request would be canceled due to the 1st error, i.e., there are no waiting requests after the error and it is proper to wait for a (new) trigger.

Reply:
I agree that the requests would be cancelled, but if there are no more requests on that five tuple, we'd leak the goroutines. The read deadline won't save us as we never attempt to read again.

Reply:
`receiverTrigger` may beed to be buffered with capacity of one, so that the trigger is not lost if sent (just) before the receive loop is read to receive it?

Reply:
I'm pretty convinced now that there is a race as you noted, but making `receiverTrigger` buffered with capacity `1` would address it.

Reply:
Pushed an update to the above effect and some more commentary.

Reply:
Doesn't that defeat the purpose of the trigger, in a way? If we buffer the trigger signal, then the recv goroutine will re-read on error even if we don't really want it to.
But it'll reread only once, which I suppose is fine?

Reply:
re-reading once would not count as busy looping, so no problem?

Reply:
Isn't there a problem still, with the buffered channel? Let's say we have the following ordering in a two request scenario:

0. Receiver is started, waits on trigger.
1. Handler receives two requests, and triggers twice. The second trigger is potentially ignored, as the channel contains a trigger already. (Crucially, I don't think there is a guarantee that the runtime will complete the receiver's read on the channel before the second send is attempted)
2. The receiver starts reading on being triggered, and errors (say, `ErrBufTooShort`). It returns the error via `clientResponse`. It now waits on another trigger before reading.
4. The handler receives the client response, and cancels the waiters. It waits for requests or responses.

Aren't we now waiting indefinitely, again?

Reply:
No, as when both (all) of the requests get a reponse (or an error), they will release their reference to the shared client, when:
- `client.close()` is called
- `c.requests` is closed, which wakes up the handler, which then exits and
  - closes the connection
  - closes `receiverTrigger` channel
  - drains the responses, which allows the receive loop to proceed in case it was blocked on sending a response
- The receive loop will then discover the closed trigger channel, which was closed before draining
- If the receive loop was blocked on `RecvMsg`, then the connection closing will wake it up instead.

Reply:
Yep, you're absolutely right - I keep missing the larger context because I doesn't all fit in my head.

Reply:
The only possibility for the receive loop to hang would be for it to get blocked on sending of a response regardless the connection being closed, trigger channel being closed, and the responses being drained once, maybe if it tries to send to the responses channel right after the handler has just drained the responses, i.e.:
- receive loop is receiving
- receive loop gets a valid response
- handler closes the connection & trigger
- handler drains the responses channel and exits
- receive loop writes on the responses channel and gets blocked?

Reply:
If this is a possible scenario, then making the `responses` channel buffered (with capacity of 1) would solve this, IMO, as then the receive loop could always send one response after the drain has happened, and would be bound to find the closed connection before sending another response?

Reply:
I don't think that's possible. The for range draining of the responses channel will block until it is closed, which only happens after the last send has occurred, hence safe without buffering.